### ISSUE TITLE
Classification on different datasets

### ISSUE THREAD

**Issue by bplevin36** [coreylowman/dfdx#77]
I'd like to use dfdx in a classification tool I'm writing.  The problem is that I'd like the tool to work across different datasets that might have different numbers of labels.  I need the final linear layer and softmax to be dynamically sized.  Is there some workaround possible, or is my application simply outside the scope of this library?

**Comment by coreylowman** [comment]
Is exposing the final layer size as a const parameter sufficient for you? Like so:
```rust
type Mlp<const N: usize> = (
    (Linear<10, 32>, ReLU),
    (Linear<32, 32>, ReLU),
    Linear<32, N>,
);

fn main() {
    let binary_mlp: Mlp<2> = Default::default();
    let mnist_mlp: Mlp<10> = Default::default();
}
```

---
Other details likely not as relevant now:

After rust stabilizes generic const exprs features you could even do something like:
```rust
trait Dataset {
    const NUM_CLASSES: usize;

    ...
}

fn mlp_for<D: Dataset>() -> Mlp<{ D::NUM_CLASSES }> { ... }
```

However this dataset is not possible now because you can't use a const associated type as a const generic parameter on stable rust.

In the mean time you'd have to have the number of classes as a generic parameter of the dataset, which is kind of unwieldy:
```rust
trait Dataset<const N: usize> {
    const NUM_LABELS: usize = N;
}

fn mlp_for<const N: usize, D: Dataset<N>>() -> Mlp<N> {
    todo!();
}
```

**Comment by bplevin36** [comment]
Thanks for the quick reply!  Unfortunately, I don't think that will work.  I want to ship a single binary that can train one type of model, but for arbitrary datasets (with an arbitrary number of classes).  This is easy with Tensorflow or Pytorch, but both of those require dynamically linking huge libraries.  The embeddability of Rust and the clarity of `dfdx` was the big draw here.  But it sounds like my concept is incompatible with it :(

**Comment by coreylowman** [comment]
Ahh I see, thanks for the details! Yeah at this time dfdx is focusing on use cases where you know pretty much everything at compile time. I do think there is a lot of flexibility with that (e.g. my example above), but it's very limiting when things are only known at run time.

Closing this for now, reopen if you have any other questions!
